src/components/data/data-utils.js

```diff
@@ -711,64 +770,22 @@ const AttributeInput = ({ autoFocus = false, value: attributeValue, initialValue
     }
   }, [attributeValue, isList])
 
-  const typeOptions = [
-    { type: 'string' },
-    { type: 'reference', tooltip: 'A link to another entity' },
-    { type: 'number' },
-    { type: 'boolean' }
-  ]
-
-  if (attributeType === 'json' || showJsonTypeOption) {
-    typeOptions.push({ type: 'json', label: 'JSON' })
-  }
-
   return h(Fragment, [
-    div({ style: { marginBottom: '1rem' } }, [
-      fieldset({ style: { border: 'none', margin: 0, padding: 0 } }, [
-        legend({ style: { marginBottom: '0.5rem' } }, [isList ? 'List item type:' : 'Type:']),
-        div({
-          style: {
-            display: 'flex', flexFlow: 'row', justifyContent: 'space-between',
-            marginBottom: '0.5rem'
-          }
-        }, _.map(({ label, type, tooltip }) => h(TooltipTrigger, { content: tooltip }, [
-          span({ style: { display: 'inline-block', whiteSpace: 'nowrap' } }, [
-            h(RadioButton, {
-              text: label || _.startCase(type),
-              name: 'edit-type',
-              checked: attributeType === type,
-              onChange: () => {
-                const newAttributeValue = convertAttributeValue(
-                  initialValue && !edited ? initialValue : attributeValue,
-                  type,
-                  defaultReferenceEntityType
-                )
-                onChange(newAttributeValue)
-              },
-              labelStyle: { paddingLeft: '0.5rem' }
-            })
-          ])
-        ]),
-        typeOptions)
+    h(AttributeTypeInput, {
+      label: isList ? 'List item type' : 'Type',
+      value: { type: attributeType, entityType: referenceEntityType },
+      entityTypes,
+      defaultReferenceEntityType: referenceEntityType,
+      showJsonTypeOption: attributeType === 'json' || showJsonTypeOption,
+      onChange: ({ type: newType, entityType: newEntityType }) => {
+        const newAttributeValue = convertAttributeValue(
+          initialValue && !edited ? initialValue : attributeValue,
+          newType,
+          newEntityType
         )
-      ]),
-      attributeType === 'reference' && div({ style: { marginTop: '0.5rem' } }, [
-        h(IdContainer, [id => h(Fragment, [
-          label({ htmlFor: id, style: { marginBottom: '0.5rem' } }, 'Referenced entity type:'),
-          h(Select, {
-            id,
-            value: defaultReferenceEntityType,
-            options: entityTypes,
-            onChange: ({ value }) => {
-              const newAttributeValue = isList ?
-                _.update('items', _.map(_.set('entityType', value)), attributeValue) :
-                _.set('entityType', value, attributeValue)
-              onChange(newAttributeValue)
-            }
-          })
-        ])])
-      ])
-    ]),
+        onChange(newAttributeValue)
+      }
+    }),
```

Comment:
The length of this file compels me to wonder if this is worth a follow-up PR that pulls some code out into a new file.

Reply:
Definitely. I'm all for breaking up the `*-utils` modules into more cohesive pieces.